modules/heat_conduction/src/userobjects/SelfShadowSideUserObject.C

```diff
@@ -0,0 +1,412 @@
+  params.addRequiredRangeCheckedParam<std::vector<PostprocessorName>>(
+      "illumination_flux",
+      "illumination_flux_size>0 && illumination_flux_size<=3",
+      "Radiation direction vector. Each component of the vector can be a constant number or a "
+      "postprocessor name (the latter enables time varying radiation directions - spatial "
+      "variation is not supported)");
+  return params;
+}
+
+SelfShadowSideUserObject::SelfShadowSideUserObject(const InputParameters & parameters)
+  : SideUserObject(parameters),
+    _dim(_mesh.dimension()),
+    _raw_direction(coupledPostprocessors("illumination_flux"))
+{
+  // we should check the coordinate system (i.e. permit only 0,0,+-1 for RZ)
+
+  // check problem dimension
+  if (_dim != 2 && _dim != 3)
+    mooseError("SelfShadowSideUserObject works only for 2 and 3 dimensional problems.");
+
+  // fetch raw direction
+  for (const auto i : index_range(_raw_direction))
+    _raw_direction[i] = &getPostprocessorValue("illumination_flux", i);
+}
+
+void
+SelfShadowSideUserObject::initialize()
+{
+  // delete list of collected lines or triangles
+  _lines.clear();
+  _triangles.clear();
+
+  // delete local qps
+  _local_qps.clear();
+
+  // update direction and rotation matrix
+  RealVectorValue direction;
+  for (const auto i : index_range(_raw_direction))
+    direction(i) = *_raw_direction[i];
+  _rotation =
+      _dim == 2 ? RotationMatrix::rotVec2DToX(direction) : RotationMatrix::rotVecToZ(direction);
+}
+
+void
+SelfShadowSideUserObject::execute()
+{
+  const SideIDType id(_current_elem->id(), _current_side);
+
+  // add triangulated sides to list
+  if (_dim == 2)
+    addLines(id);
+  else
+    addTriangles(id);
+
+  // save off rotated QP coordinates got local sides
+  auto & qps = _local_qps[id];
+  qps = _q_point;
+  rotate(qps);
+}
+
+void
+SelfShadowSideUserObject::threadJoin(const UserObject & y)
+{
+  const SelfShadowSideUserObject & uo = static_cast<const SelfShadowSideUserObject &>(y);
+
+  // merge lists
+  _lines.insert(_lines.begin(), uo._lines.begin(), uo._lines.end());
+  _triangles.insert(_triangles.begin(), uo._triangles.begin(), uo._triangles.end());
+}
+
+int
+SelfShadowSideUserObject::illumination(const SideIDType & id) const
+{
+  const auto it = _illumination_status.find(id);
+  if (it == _illumination_status.end())
+    mooseError("Illumination status was not calculated for current side.");
+  return it->second;
+}
+
+void
+SelfShadowSideUserObject::finalize()
+{
+  // rotate triangulations
+  if (_dim == 2)
+    for (auto & line : _lines)
+      rotate(line.node);
+  else
+    for (auto & triangle : _triangles)
+      rotate(triangle.node);
+
+  // [compute local projected bounding box (in x or xy)]
+
+  // communicate
+  if (_dim == 2)
+    _communicator.allgather(_lines, /*identical_buffer_sizes=*/false);
+  else
+    _communicator.allgather(_triangles, /*identical_buffer_sizes=*/false);
```

Comment:
This should be good enough performance-wise?

Reply:
for now it is. I have a way that scales better developed in the RadialAverage, but that would make the communication code _a lot_ more complicated.

Reply:
In addition a quad tree could be used for looking up intersection candidates. But again, it works in the present state and additional complexity can be added later.